In [100]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

# Tidy-Data exercises

## 1. Attendence Data 
    Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [2]:
df_att = pd.read_csv('untidy-data/attendance.csv')

In [3]:
df_att

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [7]:
# melt the table sorted by unnamed to be able to codify the tardiness column
melt_att = df_att.melt(id_vars=['Unnamed: 0'])

Unnamed: 0    variable value
0       Sally  2018-01-01     P
1        Jane  2018-01-01     A
2       Billy  2018-01-01     A
3        John  2018-01-01     P
4       Sally  2018-01-02     T
5        Jane  2018-01-02     P
6       Billy  2018-01-02     T
7        John  2018-01-02     T
8       Sally  2018-01-03     T
9        Jane  2018-01-03     T
10      Billy  2018-01-03     A
11       John  2018-01-03     H
12      Sally  2018-01-04     H
13       Jane  2018-01-04     T
14      Billy  2018-01-04     A
15       John  2018-01-04     P
16      Sally  2018-01-05     P
17       Jane  2018-01-05     T
18      Billy  2018-01-05     H
19       John  2018-01-05     P
20      Sally  2018-01-06     A
21       Jane  2018-01-06     T
22      Billy  2018-01-06     T
23       John  2018-01-06     T
24      Sally  2018-01-07     T
25       Jane  2018-01-07     A
26      Billy  2018-01-07     P
27       John  2018-01-07     P
28      Sally  2018-01-08     T
29       Jane  2018-01-08     T
30      Billy  2018-01-08     T
31       John  2018-01-08     P

In [92]:
# use a for loop to create a list of coded values for tardiness
code = []
for x in melt_att['value']:
    if x == 'P':
        code.append(0)
    elif x == 'A':
        code.append(1)
    elif x == 'T':
        code.append(.1)
    else:
        code.append(.5)
# then pass the list to the dataframe as a new column
melt_att['code'] = code

In [21]:
# Verify it worked with a .head limiter
melt_att.head(3)

Unnamed: 0    variable value  code
0      Sally  2018-01-01     P   0.0
1       Jane  2018-01-01     A   1.0
2      Billy  2018-01-01     A   1.0

In [28]:
# take a pivot table that aggregates the mean coded scores indexed by name. 
piv_att = melt_att.pivot_table(index = ['Unnamed: 0'], aggfunc = 'mean', values = 'code').reset_index()
# due to how I coded the numbers, attendence percentage was calculated by taking thier tardiness scores and subtracting it from 1
# my code essentially finds thier tardiness percentage, so the inverse of it is actually their attendance percentage
piv_att['code']= 1- piv_att['code']
# rename the nonsense column names for readablilty
piv_att.rename(columns = {'Unnamed: 0': 'Name', 'code':'tardy_rate'})

Name  tardy_rate
0  Billy      0.5250
1   Jane      0.6875
2   John      0.9125
3  Sally      0.7625

## 2. Coffee Levels
    a. Read the coffee_levels.csv file

In [93]:
df_coffee = pd.read_csv('untidy-data/coffee_levels.csv')
df_coffee

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

    b. Transform the data so that each carafe is in it's own column.

In [94]:
df_coffee = df_coffee.pivot_table(index = 'hour', columns = df_coffee.coffee_carafe)
df_coffee

coffee_amount                    
coffee_carafe             x         y         z
hour                                           
8                  0.816164  0.189297  0.999264
9                  0.451018  0.521502  0.915990
10                 0.843279  0.023163  0.144928
11                 0.335533  0.235529  0.311495
12                 0.898291  0.017009  0.771947
13                 0.310711  0.997464  0.398520
14                 0.507288  0.058361  0.864464
15                 0.215043  0.144644  0.436364
16                 0.183891  0.544676  0.280621
17                 0.391560  0.594126  0.436677

    c. is this the best shape for the data?

It is a pretty good shape for the data. it is seperated by carafe and shows the amount of cofee over time. 

## 3. Cake Recipes
    a. read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures

In [95]:
df_cake = pd.read_csv('untidy-data/cake_recipes.csv')
df_cake

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

    b. Tidy the data as necessary

In [96]:
# split the recipe:position column into two different columns, and add them to the dataframe
df_cake[['recipe', 'oven_position']]=df_cake['recipe:position'].str.split(':',expand = True)

In [97]:
# drop the recipe:position column since we dont need it anymore
df_cake = df_cake.drop(columns = 'recipe:position')

In [98]:
df_cake

225        250        275        300 recipe oven_position
0  61.738655  53.912627  74.414730  98.786784      a        bottom
1  51.709751  52.009735  68.576858  50.228470      a           top
2  57.095320  61.904369  61.196980  99.248541      b        bottom
3  82.455004  95.224151  98.594881  58.169349      b           top
4  96.470207  52.001358  92.893227  65.473084      c        bottom
5  71.306308  82.795477  92.098049  53.960273      c           top
6  52.799753  58.670419  51.747686  56.183110      d        bottom
7  96.873178  76.101363  59.571620  50.971626      d           top

    c. Which recipe is the best on average?

In [67]:
# pivot table by recipe as columns, take aggregate mean
cake_piv = df_cake.pivot_table(columns = 'recipe', aggfunc = 'mean')
cake_piv

recipe          a          b          c          d
225     56.724203  69.775162  83.888258  74.836465
250     52.961181  78.564260  67.398417  67.385891
275     71.495794  79.895930  92.495638  55.659653
300     74.507627  78.708945  59.716678  53.577368

In [99]:
# use .describe to find the averages of each column
cake_piv.describe()

recipe          a          b          c          d
count    4.000000   4.000000   4.000000   4.000000
mean    63.922201  76.736074  75.874748  62.864844
std     10.667172   4.678799  14.982231  10.032141
min     52.961181  69.775162  59.716678  53.577368
25%     55.783448  76.366985  65.477983  55.139082
50%     64.109999  78.636602  75.643338  61.522772
75%     72.248752  79.005691  86.040103  69.248535
max     74.507627  79.895930  92.495638  74.836465

Recipe b has the highest avg score

    d. Which oven temperature, on average, produces the best results?

In [101]:
# use .describe() on the original dataframe since the oven temperatures are already seperated out. 
df_cake.describe()

225        250        275        300
count   8.000000   8.000000   8.000000   8.000000
mean   71.306022  66.577437  74.886754  66.627655
std    18.636331  16.229603  17.647154  20.545432
min    51.709751  52.001358  51.747686  50.228470
25%    56.021428  53.436904  60.790640  53.213111
50%    66.522482  60.287394  71.495794  57.176230
75%    85.958804  77.774891  92.296844  73.801509
max    96.873178  95.224151  98.594881  99.248541

275 produces the best results with a mean tastiness score of 74.9

    e. Which combination of recipe, rack position, and temperature gives the best result?

In [88]:
# melt the og dataframe and sort by recipe and oven position. 
melt_cake = df_cake.melt(id_vars=['recipe','oven_position'], var_name='temperature', value_name='tastiness_score')
melt_cake

recipe oven_position temperature  tastiness_score
0       a        bottom         225        61.738655
1       a           top         225        51.709751
2       b        bottom         225        57.095320
3       b           top         225        82.455004
4       c        bottom         225        96.470207
5       c           top         225        71.306308
6       d        bottom         225        52.799753
7       d           top         225        96.873178
8       a        bottom         250        53.912627
9       a           top         250        52.009735
10      b        bottom         250        61.904369
11      b           top         250        95.224151
12      c        bottom         250        52.001358
13      c           top         250        82.795477
14      d        bottom         250        58.670419
15      d           top         250        76.101363
16      a        bottom         275        74.414730
17      a           top         275        68.576858
18      b        bottom         275        61.196980
19      b           top         275        98.594881
20      c        bottom         275        92.893227
21      c           top         275        92.098049
22      d        bottom         275        51.747686
23      d           top         275        59.571620
24      a        bottom         300        98.786784
25      a           top         300        50.228470
26      b        bottom         300        99.248541
27      b           top         300        58.169349
28      c        bottom         300        65.473084
29      c           top         300        53.960273
30      d        bottom         300        56.183110
31      d           top         300        50.971626

In [90]:
# create a mask that finds the row with the max tastiness score
melt_cake[melt_cake['tastiness_score'] == melt_cake['tastiness_score'].max()]

recipe oven_position temperature  tastiness_score
26      b        bottom         300        99.248541